draft-ietf-mls-protocol.md

```diff
@@ -1353,66 +1353,131 @@ Given these inputs, the derivation of secrets for an epoch
-to synchronize the injection of a PSK into the MLS key schedule.
+At any epoch, including the initial state, an application can decide to
+synchronize the injection of one or more PSKs into the MLS key schedule.
 
-This mechanism can be used to improve security in the cases where
-having a full run of updates across members is too expensive or in
-the case where the external group key establishment mechanism provides
+The injecting of an external PSK (EPSK) can improve security in the cases
+where having a full run of updates across members is too expensive or in the
+case where the external group key establishment mechanism provides
 stronger security against classical or quantum adversaries.
 
 The security level associated with the PSK injected in the key schedule
 SHOULD match at least the security level of the ciphersuite in use in
 the group.
 
-Note that, as a PSK may have a different lifetime than an update, it
-does not necessarily provide the same FS or PCS guarantees than
-a Commit message.
+[[Open ISSUE: Define "security level", and what it means to match the
+security level of the ciphersuite used in the group.]]
+
+Note that, as a PSK may have a different lifetime than an update, it does
+not necessarily provide the same FS or PCS guarantees as a Commit
+message.
+
+[[OPEN ISSUE: Clarify lifetime vs security level mandated above. E.g. if
+the PSK security expires before the next update (shorter PSK lifetime than
+update), does that constitute a weaker security level]]
 
 [[OPEN ISSUE: We have to decide if we want an external coordination
 via the application of a Handshake proposal.]]
 
+A PSK may also be used within MLS in the following cases:
+
+  - Re-Initialization: If, during the lifetime of the group, a change in the
+    fixed group parameters becomes necessary, e.g. if the ciphersuite used
+    by the group is deprecated or if the protocol version should be upgraded,
+    a PSK can be used to reboot the group with the desired parameters.
+
+  - Recovery: If the group state of one or more members of the group deviates
+    from the rest, they can be re-added to the group using a `recovery_key` from
+    a previously known shared group state.
+
+  - Branching: A PSK may be used to bootstrap a subset of current group
+    members into a new group. This applies if a subset of current group
+    members wish to branch based on the current group state.
+
+Regardless of its origin, the PSK injected in the key schedule must be unique.
+We ensure this by deriving it using a nonce and a `PSKLabel` as follows.
+
+~~~~~
+enum {
+  re-initialization(0),
+  branch(1),
+  recovery(2),
+  external(3),
+  (255)
+} PSKType;
+
+PSKLabel(psktype) =
+  select (psktype) {
+    case re-initialization: "re-inizialization";
+    case branch: "branch";
+    case recovery: "recovery";
+    case external: "external";
+}
+
+Derived_PSK =
+  Derive-Secret(PSK, Hash(PSKLabel(psktype) || psk_nonce))
```

Comment:
Over in TLS, we worked on an interface for injecting PSKs:

   https://tools.ietf.org/html/draft-ietf-tls-external-psk-importer-05

We concluded binding the imported PSK to the target protocol and KDF was sufficient. Maybe the same thing should be done here, i.e., to ensure that an injected PSK for MLS v1.0 does not collide with one injected for a different version?

Reply:
Adding the MLS version number to the PSK type sounds like a solid addition.